spec/unit/reports/http_spec.rb

```diff
@@ -1,101 +1,117 @@
-        'testing', 8080, use_ssl: false, ssl_context: nil
-      ).and_return(http)
+        anything, anything, use_ssl: false, ssl_context: nil
+      ).and_call_original
 
       subject.process
     end
   end
 
   describe "when making a request" do
-    let(:connection) { double("connection") }
-    let(:httpok) { Net::HTTPOK.new('1.1', 200, '') }
-    let(:options) { {:metric_id => [:puppet, :report, :http]} }
+    it "uses the path specified by the 'reporturl' setting" do
+      req = stub_request(:post, url)
+
+      subject.process
 
-    before :each do
-      expect(Puppet::Network::HttpPool).to receive(:connection).and_return(connection)
+      expect(req).to have_been_requested
     end
 
-    it "should use the path specified by the 'reporturl' setting" do
-      report_path = URI.parse(Puppet[:reporturl]).path
-      expect(connection).to receive(:post).with(report_path, anything, anything, options).and_return(httpok)
+    it "uses the username and password specified by the 'reporturl' setting" do
+      Puppet[:reporturl] = "https://user:pass@puppet.example.com/report/upload"
+
+      req = stub_request(:post, %r{/report/upload}).with(basic_auth: ['user', 'pass'])
+
+      subject.process
+
+      expect(req).to have_been_requested
+    end
+
+    it "passes metric_id options" do
+      stub_request(:post, url)
+
+      expect_any_instance_of(Puppet::Network::HTTP::Connection).to receive(:post).with(anything, anything, anything, hash_including(metric_id: [:puppet, :report, :http])).and_call_original
 
       subject.process
     end
 
-    it "should use the username and password specified by the 'reporturl' setting" do
-      Puppet[:reporturl] = "https://user:pass@myhost.mydomain:1234/report/upload"
+    it "passes the report as YAML" do
+      req = stub_request(:post, url).with(body: subject.to_yaml)
 
-      expect(connection).to receive(:post).with(anything, anything, anything,
-                                                {:metric_id => [:puppet, :report, :http],
-                                                 :basic_auth => {
-                                                   :user => 'user',
-                                                   :password => 'pass'
-                                                 }}).and_return(httpok)
+      subject.process
+
+      expect(req).to have_been_requested
+    end
+
+    it "sets content-type to 'application/x-yaml'" do
+      req = stub_request(:post, url).with(headers: {'Content-Type' => 'application/x-yaml'})
 
       subject.process
+
+      expect(req).to have_been_requested
     end
 
-    it "should give the body as the report as YAML" do
-      expect(connection).to receive(:post).with(anything, subject.to_yaml, anything, options).and_return(httpok)
+    it "doesn't log anything if the request succeeds" do
+      stub_request(:post, url).to_return(status: [200, "OK"])
 
       subject.process
+
+      expect(@logs).to eq([])
```

Comment:
this would be one that probably benefit from ensuring the request is actually requested

Reply:
Updated